detach/detach.go

```diff
@@ -0,0 +1,97 @@
+	// src could be closed or otherwise cleaned up, the wrapper should likely create new values and clean those up rather
+	// than just copying the values from src to dst.
+	Async func(dst, src context.Context, next func(ctx context.Context))
+}
+
+type hooksRegistry struct {
+	mu    sync.RWMutex
+	hooks []Hooks
+}
+
+func (r *hooksRegistry) add(hooks Hooks) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	r.hooks = append(r.hooks, hooks)
+}
+
+func (r *hooksRegistry) allHooks() []Hooks {
+	r.mu.RLock()
+	defer r.mu.RUnlock()
+
+	return r.hooks
+}
+
+var registry = new(hooksRegistry)
+
+// Register adds a new Hooks to the library's registry. If all functions in the Hooks are nil, Register will panic.
+func Register(hooks Hooks) {
+	if hooks.Inline == nil && hooks.Async == nil {
+		panic("cannot register Hooks with all nil functions")
+	}
+	registry.add(hooks)
+}
+
+// Inline returns a new context who inherits the values attached using the Inline Hooks registered with the library
+// from the parent but replaces the parent's timeout/cancel with the timeout given and the returned cancel func.
+//
+// Inline should be used when you need to run something within the current call, rather than in a new goroutine, but
+// need to ignore any timeouts/cancellations from the parent context, such as cleaning up after a call that might have
+// been timed out or canceled.
+func Inline(parent context.Context, timeout time.Duration) (context.Context, context.CancelFunc) {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	for _, hooks := range registry.allHooks() {
+		if hooks.Inline != nil {
+			ctx = hooks.Inline(ctx, parent)
+		}
+	}
+	return ctx, cancel
+}
+
+// Async wraps run in the Async Hooks registered with the library, creates a new context with the given timeout, and
+// calls the wrapped run function with that context. The wrappers will attach values from the parent context to the new
+// context and perform any cleanup tasks after run completes. The new context will be canceled when Async returns.
+//
+// Async should be used when you want to run an async task as a part of another call, but need to keep some values from
+// the parent context.
+func Async(parent context.Context, timeout time.Duration, run func(ctx context.Context)) {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	allHooks := registry.allHooks()
+	for i := len(allHooks) - 1; i >= 0; i-- {
+		hooks := allHooks[i]
+		if hooks.Async != nil {
+			// assign 'next' to an inner variable rather than using 'run! directly in the wrapped call to avoid an infinite
+			// loop.
+			next := run
+			run = func(ctx context.Context) {
+				// if you just used 'run' in here, it would always use the last value of "run", so it would call itself over and
+				// over again, creating an infinite loop.
+				hooks.Async(ctx, parent, next)
+			}
+		}
+	}
+	run(ctx)
```

Comment:
🔕 should this be `go run(ctx)` instead? without that the caller would need to do `go detach.Async(...)`, which sounds redundant (with both `go` and `Async`). but I also don't feel strongly either way.

Reply:
I guess it makes sense for a use case that this is not the whole background goroutine, just a portion of it, e.g.:
```go
go func() {
  // some preparing work
  detach.Async(...)
  // some cleanup work
}
```

so ignore my previous comment 😅